frontend/src/lib/components/modals/AddProjectUser.svelte

```diff
@@ -2,70 +2,69 @@
 	import Form from '$lib/forms/Form.svelte';
 	import Input from '$lib/forms/Input.svelte';
 	import t from '$lib/i18n';
-	import Modal, { CloseReason } from './Modal.svelte';
+	import Modal, { DialogResponse } from './Modal.svelte';
 	import { addProjectUser } from '$lib/services/project.service';
 	import { ProjectRole } from '$lib/gql/graphql';
+	import { lexSuperForm, lexSuperValidate, Select } from '$lib/forms';
+	import { z } from 'zod';
+
+	const formSchema = z.object({
+		email: z.string().email($t('project_page.add_user.email_required')),
+		role: z.enum([ProjectRole.Editor, ProjectRole.Manager])
+	});
 
 	export let projectId: string;
 	let modal: Modal;
+	let {form, errors, valid, update, reset, message } = lexSuperForm(formSchema);
 	async function openModal(): Promise<void> {
-		if ((await modal.openModal()) === CloseReason.Cancel) return;
-		emailError = '';
-		error = '';
-		//invalid try again
-		if (!email) {
-			emailError = $t('project_page.add_user.email_required');
-			return await openModal();
-		}
-		const result = await addProjectUser({ projectId, userEmail: email, role });
+		if (await modal.openModal() === DialogResponse.Cancel) return;
+		//validate form, we're using a wrapper over the library method because we're doing this all client side.
+		await lexSuperValidate($form, formSchema, update);
+		// go back to the top and await a response from the modal again.
+		if (!$valid) return await openModal();
+		const result = await addProjectUser({ projectId, userEmail: $form.email, role: $form.role });
 		if (result.error) {
-			error = result.error.message;
+			$message = result.error.message;
+			// again go back to the top and await a response from the modal.
 			return await openModal();
 		}
 		modal.close();
 	}
-
-	let email: string;
-	let role: ProjectRole = ProjectRole.Editor;
-	let emailError: string | undefined;
-	let error: string;
 </script>
 
 <button class="badge badge-lg badge-success cursor-pointer" on:click={openModal}>
 	<span class="i-mdi-plus"/> {$t('project_page.add_user.add_button')}
 </button>
-<Modal bind:this={modal} on:close={() => (email = '')} onBottom>
+<Modal bind:this={modal} on:close={() => reset()} onBottom>
```

Comment:
if you like, shorthand for this is `on:close={reset}`

Reply:
actually if you do TS will warn you of a potential error. On close emits an event parameter which would then be passed in to reset if we were to do that and reset has an optional first argument which the event parameter conflicts with, which is why we need to call it this way.

Reply:
oh I see, `reset` is actually coming from superform... ok, that makes more sense.  I assumed it was a local function.